examples/ota-requestor-app/esp32/main/OTARequesterImpl.cpp

```diff
@@ -0,0 +1,279 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/device/OperationalDeviceProxy.h>
+#include <app/server/Server.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <zap-generated/CHIPClientCallbacks.h>
+#include <zap-generated/CHIPClusters.h>
+
+#include <esp_log.h>
+
+#include "BDXDownloader.h"
+#include "OTARequesterImpl.h"
+#include "OTAUpdater.h"
+
+#define TAG "OTARequesterImpl"
+
+using chip::ByteSpan;
+using chip::CharSpan;
+using chip::EndpointId;
+using chip::FabricIndex;
+using chip::NodeId;
+using chip::Optional;
+using chip::Server;
+using chip::VendorId;
+using chip::bdx::TransferSession;
+using chip::Callback::Callback;
+using chip::Inet::IPAddress;
+using chip::System::Layer;
+using chip::Transport::PeerAddress;
+using namespace chip::Messaging;
+using namespace chip::app::device;
+using namespace chip::app::Clusters::OtaSoftwareUpdateProvider::Commands;
+
+void OnQueryImageResponse(void * context, const QueryImageResponse::DecodableType & response);
+void OnQueryImageFailure(void * context, EmberAfStatus status);
+void OnApplyUpdateResponse(void * context, const ApplyUpdateResponse::DecodableType & response);
+void OnApplyUpdateRequestFailure(void * context, EmberAfStatus status);
+
+void OnConnected(void * context, OperationalDeviceProxy * operationalDeviceProxy);
+void OnConnectionFailure(void * context, OperationalDeviceProxy * operationalDeviceProxy, CHIP_ERROR error);
+
+void OnBlockReceived(void * context, const chip::bdx::TransferSession::BlockData & blockdata);
+void OnTransferComplete(void * context);
+void OnTransferFailed(void * context, BdxDownloaderErrorTypes status);
+
+enum OTARequestorCommands
+{
+    kCommandQueryImage = 0,
+    kCommandApplyUpdateRequest,
+    kCommandNotifyUpdateApplied,
+};
+
+// TODO: Encapsulate these globals and the callbacks in some class
+OperationalDeviceProxy gOperationalDeviceProxy;
```

Comment:
There is a more complete version of `OperationalDeviceProxy` that resides in src/app/OperationalDeviceProxy.h. The one that sits in src/app/device is now obsolete (I'm planning on removing it today). The object you should use now resides in Server class and you'll need to instantiate one like so https://github.com/project-chip/connectedhomeip/blob/master/examples/ota-requestor-app/linux/main.cpp#L267-L269

Reply:
Also FYI, just discovered a bug and it is being fixed here so when you make this change, you'll need keep this in mind.